refactor: Extract MIB_BYTES constant for init.cpp

Empact opened the pull request:
This is a common concept which is more communicative and explicit if articulated, IMO.

This could be applied elsewhere if there is a better place to reference it in common, but I don't see it. 

Pulled from #25315 to ease review there.

Empact committed:
refactor: Expose MIN_MIB_FOR_BLOCK_FILES, MIN_BYTES_FOR_BLOCK_FILES constant

Empact committed:
refactor: Extract MIB_BYTES constant for init.cpp

Co-authored-by: benthecarman <benthecarman@live.com>
Co-authored-by: Justin Litchfield <litch@me.com>
Co-authored-by: Liran Cohen <c.liran.c@gmail.com>
Co-authored-by: Ryan Loomba <ryan.loomba@gmail.com>
Co-authored-by: Buck Perley <bucko.perley@gmail.com>
Co-authored-by: bajjer <bajjer@bajjer.xyz>
Co-authored-by: Suhail Saqan <suhail.saqan@gmail.com>
Co-authored-by: Christopher Sweeney <sweeney.chris@gmail.com>
Co-authored-by: Alyssa <orbitalturtle@protonmail.com>
Co-authored-by: Ben Schroth <ben@styng.social>
Co-authored-by: Jason Hester <mail@jason-hester.me>
Co-authored-by: Matt Clough <Matt.clough@pm.me>
Co-authored-by: Elise Schedler <eliseschedler@gmail.com>
Co-authored-by: ghander <cen254@gmail.com>
Co-authored-by: PopeLaz <btclz@fastmail.com>

laanwj commented:
Concept ACK

>  if there is a better place to reference it in common, but I don't see it.

Right, not sure `init.cpp` is the best place for this constant, but the idea of having it as a named constant makes sense to me, it tells anyone encountering the code about intent and the unit.

brunoerg reviewed (commented):
Concept ACK

It's easier to read and maintain the code this way imo.

w0xlt reviewed (commented):
Approach ACK

Adding the unit makes the code more readable (in the same way that using `std::chrono` instead of `int` makes the code better)

MarcoFalke commented on src/init.cpp:
```diff
@@ -118,6 +118,8 @@ using node::nPruneTarget;
 static const bool DEFAULT_PROXYRANDOMIZE = true;
 static const bool DEFAULT_REST_ENABLE = false;
 
+static constexpr size_t MIB_BYTES = 1024 * 1024;
```
This will force all places where this is used to uint64_t, even if they are supposed to be signed integers. Not sure, but maybe a `#define` or `int` type is better, so that the sign of the surrounding code is chosen?

laanwj replied:
Ouch, good point. I'm starting to doubt this change now.

Empact replied:
Thanks @MarcoFalke, this was indeed surprising to me:
> Otherwise, if the unsigned operand's conversion rank is greater or equal to the conversion rank of the signed operand, the signed operand is converted to the unsigned operand's type.https://en.cppreference.com/w/cpp/language/operator_arithmetic#Conversions

Empact replied:
Though I maintain `uint64_t` for `GB_BYTES` to maintain protections against overflow in `PruneGBtoMiB`.

Empact replied:
In an effort to validate these concerns, I ran build with `-Wsign-conversion` and `-Wsign-compare`, with the new and old code, and did not see any warnings relative to `MIB_BYTES`, `MIN_PRUNE_TARGET_GB`, or `_FOR_BLOCK_FILES`. I also reviewed each impacted line and did not identify such issues.

Prior, `uint64_t` code here: https://github.com/Empact/bitcoin/commits/2022-06-mib-bytes-sign-conversion-test
Current, `int32_t` code here: https://github.com/Empact/bitcoin/commits/2022-06-mib-bytes-sign-conversion

Empact committed:
Expose MIN_PRUNE_TARGET_GB in guiconstants.h

This requires creating a non-gui analog to guiconstants.h, and
lifting GB/MIB_BYTES and MIN_*_FOR_BLOCK_FILES into it.

achow101 commented:
cc @mzumsande

Empact committed:
build: Enable -Wsign-compare and -Wsign-conversion

mzumsande commented on src/constants.h:
```diff
@@ -0,0 +1,27 @@
+static constexpr uint64_t GB_BYTES{1000000000};
+
+/* One mebibyte (MiB) in bytes */
+static constexpr int32_t MIB_BYTES{1024 * 1024};
```
How about introducing this right from the beginning? It's a bit weird that `MIB_BYTES` jumps files with each commit, from `init.cpp` to `validation.h` to its final destination `constants.h`.

mzumsande reviewed (commented):
Concept ACK.

`git grep "1024 / 1024"` shows a few more spots that could make use of the constant.

achow101 commented:
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.
